Fix: Substitution plugin API compatibility

Adnan-cds opened the pull request:
The LinkIt [substitution plugin from this module](https://github.com/localgovdrupal/localgov_page_components/blob/1.x/src/Plugin/Linkit/Substitution/ParagraphsLibraryItem.php) has become incompatible with the recently released LinkIt version 6.1.1.  This is due to API changes.  Previously, [Drupal\linkit\SubstitutionInterface::getUrl()](https://git.drupalcode.org/project/linkit/-/blob/6.0.x/src/SubstitutionInterface.php?ref_type=heads#L23) was expected to return [Drupal\Core\GeneratedUrl](https://git.drupalcode.org/project/linkit/-/blob/6.1.0/src/SubstitutionInterface.php?ref_type=tags#L20).  Now it expects [Drupal\Core\Url](https://git.drupalcode.org/project/linkit/-/blob/6.1.1/src/SubstitutionInterface.php?ref_type=tags#L20).  Given this is a breaking change, the LinkIt project should have made a version 7.0.0 release instead of version 6.1.1.

Also added a test to cover the substitution plugin where this happened.

I had to drop Drupal 9 compatibility from this module as [LinkIt](https://www.drupal.org/project/linkit) 6.1.1 is incompatible with Drupal versions prior to 10.1.  Does it mean we need to start a 3.x (or 2.x) branch for this module? @finnlewis

Adnan-cds committed:
Fix: Substitution plugin.

The LinkIt substitution plugin became incompatible with the recently release
LinkIt version 6.1.1.  This is due to API changes.  Resolved now.

Also added a test to cover the substitution plugin where this happened.

Adnan-cds committed:
Chore: Dropping Drupal 9 compatibility.

LinkIt version 6.1.1 is incompatible with Drupal 9.

Adnan-cds committed:
Fix: Removing remaining Drupal 9 checks.

andybroomfield commented:
Yes, looks like we need to provide a 3.x for Drupal 10 and lock 2.x to the 6.0.x branch, as that should be the drupal 9 version.

andybroomfield commented:
Has 6.0.x of link made the same change?
https://git.drupalcode.org/project/linkit/-/blob/6.0.x/src/SubstitutionInterface.php?ref_type=heads#L23

stephen-cox commented:
This is going to require a 1.x branch with D9 support and a 2.x branch with D10 support.

This module is required by Services which currently has a 2.x branch that support D9 and D10. One option is a new 3.x branch of Services which supports Drupal 10 only, but dropping D10 support from the 2.x branch of Services could cause all sorts of problems. Perhaps we can get round this by updating Services to require either the 1.x or 2.x versions of Page Components and letting Composer determine the correct version to install.

andybroomfield commented:
@stephen-cox I don't think that will be nessecary as the same change was made to the Drupal 9 version on Linkit. 
I think this might actully be uneeded for now as they since [added compatiabilty back in change in 6.1.2 and 6.0.2](https://git.drupalcode.org/project/linkit/-/commit/b325c7e5f47cd2a364e9629824b696e493e30432) though its marked as deprected.
We should actully be able to provide an update as long as we can gurantee linkit 6.0.2+ Or 6.1.2+ (exclude 6.1.0 and 6.1.1).

Adnan-cds committed:
Update: Restoring Drupal 9 compatibility.

The newly released LinkIt 6.0.2, which is Drupal 9 compatible, has restored API
compatibility.  So no need to drop Drupal 9 support.

Adnan-cds commented:
I have [restored](https://github.com/localgovdrupal/localgov_page_components/pull/45/commits/137b84a0918f676a69c1fc44ba813ebca9465de7) Drupal 9 compatibility :)

andybroomfield commented on composer.json:
```diff
@@ -8,7 +8,7 @@
         "drupal/entity_browser": "^2.5",
         "drupal/inline_entity_form": "^1.0-rc6",
-        "drupal/linkit": "^6.1.1",
+        "drupal/linkit": "^6.0.2",
```
May I suggest we try
```
~6.0.2 || ^6.1.2
```
This should
- [Disallow 6.0.1](https://jubianchi.github.io/semver-check/#/~6.0.2%20||%20^6.1.2/6.0.1)
- [Allow 6.0.2](https://jubianchi.github.io/semver-check/#/~6.0.2%20||%20^6.1.2/6.0.2)
- [Disallow 6.1.1](https://jubianchi.github.io/semver-check/#/~6.0.2%20||%20^6.1.2/6.1.1)
- [Allow 6.1.2](https://jubianchi.github.io/semver-check/#/~6.0.2%20||%20^6.1.2/6.1.2)
- [Allow 6.2.0](https://jubianchi.github.io/semver-check/#/~6.0.2%20||%20^6.1.2/6.2.0)

Adnan-cds replied:
Hi Andy, this is what I was thinking when restricting drupal/linkit to ^6.0.2:
- It covers anything above and including 6.0.2.
- It includes both 6.1.1 and 6.1.2.  The change from this pull request is compatible with 6.1.1 afterall.
- In the future it should also cover 6.2.x.

So perhaps ^6.0.2 is enough here?

andybroomfield replied:
ok, I think it is 6.1.0 that is the problem, but then I don't think people would install that.

Adnan-cds replied:
> it is 6.1.0 that is the problem

That's a good point.  I missed 6.1.0.  Fixed now.

Adnan-cds committed:
Fix: LinkIt dependency

This module is now incompatible with LinkIt 6.1.0.  Updating composer version strings accordingly.

andybroomfield commented on localgov_page_components.info.yml:
```diff
@@ -1,10 +1,11 @@
 type: module
 description: 'Disguise "Paragraphs library" as "Page components" for better UX.'
-core_version_requirement: ^9 || ^10
+core_version_requirement: ">=9"
```
Maybe we go back to 
```
core_version_requirement: ^9.4 || ^10
```
Unless we think this also supports Drupal 11?

andybroomfield reviewed (approved):
Aside from above comment I'm happy to approve these changes.
Tested with link and contact paragraph page component on new page using linkit to add a link.
Verifed the link points to the link in the paragraph component.

Adnan-cds commented:
Thanks for the approval.  Let's keep the `core_version_requirement` open ended for now.  This will be a good test to see if" >=9" is a good idea or not.
